### PR TITLE
Add errors for Mixed Dataloader Eval

### PR DESCRIPTION
**Pull request by rishab-partha**
# What does this PR do?

Now, if a user passes in a list of some DataLoaders/DataSpecs and Evaluators to the Trainer class, it no longer tries to apply all metrics to the DataLoaders/DataSpecs, causing a crash. Instead, it raises a ValueError and instructs the user to wrap the DataLoaders in an Evaluator beforehand to prevent unexpected behavior.

# What issue(s) does this change relate to?

[CO-1763](https://mosaicml.atlassian.net/browse/CO-1763
)
# Before submitting
- [x] Have you read the [contributor guidelines](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md)?
- [ ] Is this change a documentation change or typo fix? If so, skip the rest of this checklist.
- [ ] Was this change discussed/approved in a GitHub issue first? It is much more likely to be merged if so.
- [x] Did you update any related docs and document your change?
- [x] Did you update any related tests and add any new tests related to your change? (see [testing](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#running-tests))
- [x] Did you run the tests locally to make sure they pass?
- [x] Did you run `pre-commit` on your change? (see the `pre-commit` section of [prerequisites](https://github.com/mosaicml/composer/blob/dev/CONTRIBUTING.md#prerequisites))

[CO-1763]: https://mosaicml.atlassian.net/browse/CO-1763?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: composer/trainer/trainer.py [hunk 1]
```diff
@@ -469,12 +469,13 @@ class Trainer:
             Setting this to ``True`` will force schedulers to be stepped every batch,
             while ``False`` means schedulers stepped every epoch. ``None`` indicates the default behavior.
             (default: ``None``)
-        eval_dataloader (DataLoader | DataSpec | Evaluator | Sequence[Evaluator], optional): The :class:`.DataLoader`,
+        eval_dataloader (Iterable | DataSpec | Evaluator | Sequence[Evaluator], optional): The :class:`.Iterable`,
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:472]: Can we leave this as DataLoader (which most users will do) and add on Iterable as another option?

File: composer/trainer/trainer.py [hunk 2]
```diff
@@ -469,12 +469,13 @@ class Trainer:
...
             :class:`.DataSpec`, :class:`.Evaluator`, or sequence of evaluators for the evaluation data.
 
             To evaluate one or more specific metrics across one or more datasets, pass in an
-            :class:`.Evaluator`. If a :class:`.DataSpec` or :class:`.DataLoader` is passed in, then all
-            metrics returned by ``model.get_metrics()`` will be used during evaluation.
+            :class:`.Evaluator`. If a :class:`.DataSpec` or :class:`.Iterable` is passed in, then all
+            metrics returned by ``model.get_metrics()`` will be used during evaluation. If another class is used
+            with :class:`.Evaluator` in a list, a ValueError will be raised.
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:478]: Nit: More concise to say what to expect

```suggestion
            metrics returned by ``model.get_metrics()`` will be used during evaluation. If a :class:`.Evaluator` is specified in a list, all eval dataloaders must be :class:`.Evaluator`.

File: composer/trainer/trainer.py [hunk 3]
```diff
@@ -2588,9 +2601,10 @@ def eval(
             drop duplicate samples.
 
         Args:
-            eval_dataloader (DataLoader | DataSpec | Evaluator | Sequence[Evaluator], optional): Dataloaders
+            eval_dataloader (Iterable | DataSpec | Evaluator | Sequence[Evaluator], optional): Dataloaders
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:2604]: same comment as above on keeping dataloader

File: composer/trainer/trainer.py [hunk 4]
```diff
@@ -1144,10 +1145,16 @@ def __init__(
         else:
             eval_metrics = deepcopy(self.state.model.get_metrics(is_train=False))
             model_metric_names = [str(k) for k in eval_metrics.keys()]
+            dataloader_tupled = ensure_tuple(eval_dataloader)
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1148]: nit: can we just do this to keep it simple?
```suggestion
            eval_dataloader = ensure_tuple(eval_dataloader)
```

File: composer/trainer/trainer.py [hunk 5]
```diff
@@ -2588,9 +2601,10 @@ def eval(
             drop duplicate samples.
 
         Args:
-            eval_dataloader (DataLoader | DataSpec | Evaluator | Sequence[Evaluator], optional): Dataloaders
+            eval_dataloader (Iterable | DataSpec | Evaluator | Sequence[Evaluator], optional): Dataloaders
                 for evaluation.  If not provided, defaults to using the
-                ``eval_dataloader`` provided to the trainer init().
+                ``eval_dataloader`` provided to the trainer init(). If a list of Evaluators is provided
+                mixed with other classes, a ValueError will be raised.
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:2607]: I think it's fine to not duplicate here

```suggestion
                ``eval_dataloader`` provided to the trainer init().
```

File: composer/trainer/trainer.py [hunk 6]
```diff
@@ -2600,9 +2614,16 @@ def eval(
             eval_metrics = deepcopy(self._original_model.get_metrics(is_train=False))
             metric_names = [str(k) for k in eval_metrics.keys()]
 
+            dataloader_tupled = ensure_tuple(eval_dataloader)
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:2617]: same nit as above for keeping same name

**Review by mvpatel2000 — COMMENTED** [review]
This looks great! Mostly minor cleanup comments. Can you please also add a unit test verifying this behavior? 

Also, does `eval()` need the same logic added to init and fit?

**Comment by rishab-partha** [comment]
fixed all above

File: composer/trainer/trainer.py [hunk 7]
```diff
@@ -1144,10 +1145,16 @@ def __init__(
         else:
             eval_metrics = deepcopy(self.state.model.get_metrics(is_train=False))
             model_metric_names = [str(k) for k in eval_metrics.keys()]
+            eval_dataloader = ensure_tuple(eval_dataloader)
+
+            evaluator_types = [isinstance(evaluator, Evaluator) for evaluator in eval_dataloader]
+
+            if any(evaluator_types) and not all(evaluator_types):
+                raise ValueError('Mixing Evaluator with other classes is not allowed, please wrap'
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1153]: nit
```suggestion
            evaluator_types = [isinstance(evaluator, Evaluator) for evaluator in eval_dataloader]
            if any(evaluator_types) and not all(evaluator_types):
                raise ValueError('Mixing Evaluator with other classes is not allowed, please wrap'
```

File: composer/trainer/trainer.py [hunk 8]
```diff
@@ -1710,10 +1717,16 @@ def fit(
             # could be DDP / DeepSpeed wrapped.
             eval_metrics = self._original_model.get_metrics(is_train=False)
             metric_names = [str(k) for k in eval_metrics.keys()]
+            eval_dataloader = ensure_tuple(eval_dataloader)
+
+            evaluator_types = [isinstance(evaluator, Evaluator) for evaluator in eval_dataloader]
+
+            if any(evaluator_types) and not all(evaluator_types):
+                raise ValueError('Mixing Evaluator with other classes is not allowed, please wrap'
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:1725]: nit:
```suggestion
            evaluator_types = [isinstance(evaluator, Evaluator) for evaluator in eval_dataloader]
            if any(evaluator_types) and not all(evaluator_types):
                raise ValueError('Mixing Evaluator with other classes is not allowed, please wrap'
```

File: composer/trainer/trainer.py [hunk 9]
```diff
@@ -2600,9 +2613,16 @@ def eval(
             eval_metrics = deepcopy(self._original_model.get_metrics(is_train=False))
             metric_names = [str(k) for k in eval_metrics.keys()]
 
+            eval_dataloader = ensure_tuple(eval_dataloader)
+
+            evaluator_types = [isinstance(evaluator, Evaluator) for evaluator in eval_dataloader]
+
+            if any(evaluator_types) and not all(evaluator_types):
+                raise ValueError('Mixing Evaluator with other classes is not allowed, please wrap'
```
Comment (mvpatel2000) [inline composer/trainer/trainer.py:2621]:
```suggestion
            evaluator_types = [isinstance(evaluator, Evaluator) for evaluator in eval_dataloader]
            if any(evaluator_types) and not all(evaluator_types):
                raise ValueError('Mixing Evaluator with other classes is not allowed, please wrap'
```

**Review by mvpatel2000 — COMMENTED** [review]
LGTM! Will wait for review from @dakinggg as well though before approval

File: composer/trainer/trainer.py [hunk 10]
```diff
@@ -479,12 +479,13 @@ class Trainer:
             Setting this to ``True`` will force schedulers to be stepped every batch,
             while ``False`` means schedulers stepped every epoch. ``None`` indicates the default behavior.
             (default: ``None``)
-        eval_dataloader (DataLoader | DataSpec | Evaluator | Sequence[Evaluator], optional): The :class:`.DataLoader`,
+        eval_dataloader (Iterable | DataLoader | DataSpec | Evaluator | Sequence[Evaluator], optional): The :class:`.Iterable`,
```
Comment (dakinggg) [inline composer/trainer/trainer.py:482]: this docstring should still mention `DataLoader`, right?

**Review by dakinggg — COMMENTED** [review]
lgtm, a couple nits

File: composer/trainer/trainer.py [hunk 11]
```diff
@@ -1155,10 +1156,15 @@ def __init__(
         else:
             eval_metrics = deepcopy(self.state.model.get_metrics(is_train=False))
             model_metric_names = [str(k) for k in eval_metrics.keys()]
+            eval_dataloader = ensure_tuple(eval_dataloader)
+
+            evaluator_types = [isinstance(evaluator, Evaluator) for evaluator in eval_dataloader]
+            if any(evaluator_types) and not all(evaluator_types):
+                raise ValueError('Mixing Evaluator with other classes is not allowed, please wrap'
```
[the remark this answers is not in the document]
Comment (rishab-partha) [reply composer/trainer/trainer.py:1163]: added

**Comment by rishab-partha** [comment]
fixed nits!

File: composer/trainer/trainer.py [hunk 12]
```diff
@@ -1726,10 +1733,15 @@ def fit(
             # could be DDP / DeepSpeed wrapped.
             eval_metrics = self._original_model.get_metrics(is_train=False)
             metric_names = [str(k) for k in eval_metrics.keys()]
+            eval_dataloader = ensure_tuple(eval_dataloader)
+
+            evaluator_types = [isinstance(evaluator, Evaluator) for evaluator in eval_dataloader]
+            if any(evaluator_types) and not all(evaluator_types):
+                raise ValueError('Mixing Evaluator with other classes is not allowed, please wrap'
```
Comment (dakinggg) [inline composer/trainer/trainer.py:1740]: same requested change to add the types, sorry, missed it the first time

File: composer/trainer/trainer.py [hunk 13]
```diff
@@ -2617,9 +2629,15 @@ def eval(
             eval_metrics = deepcopy(self._original_model.get_metrics(is_train=False))
             metric_names = [str(k) for k in eval_metrics.keys()]
 
+            eval_dataloader = ensure_tuple(eval_dataloader)
+
+            evaluator_types = [isinstance(evaluator, Evaluator) for evaluator in eval_dataloader]
+            if any(evaluator_types) and not all(evaluator_types):
+                raise ValueError('Mixing Evaluator with other classes is not allowed, please wrap'
```
Comment (dakinggg) [inline composer/trainer/trainer.py:2636]: same

**Review by dakinggg — APPROVED** [review]

**Pull request merged**